subsys/lorawan/services/multicast.c

```diff
@@ -0,0 +1,373 @@
+		 * of frames
+		 */
+		uint32_t fcnt_max;
+	} mc_group;
+	/** Start of the Class C window as GPS epoch modulo 2^32 */
+	uint32_t session_time;
+	/** Maximum duration of MC session before device reverts to class A */
+	uint32_t session_timeout;
+	/** Receive parameters for MC session */
+	McRxParams_t rx_params;
+
+	struct k_work_delayable session_start_work;
+	struct k_work_delayable session_stop_work;
+};
+
+static struct multicast_context ctx[LORAMAC_MAX_MC_CTX];
+
+static void multicast_settings_store(uint8_t group_id)
+{
+	char path[30];
+
+	snprintf(path, sizeof(path), "%s/groups/%u", MULTICAST_SETTINGS_BASE, group_id);
+
+	settings_save_one(path, &ctx[group_id].mc_group, sizeof(ctx[group_id].mc_group));
+}
+
+static int multicast_settings_load(const char *name, size_t len,
+				   settings_read_cb read_cb, void *cb_arg)
+{
+	const char *remainder;
+	int rc, group_id;
+
+	if (settings_name_steq(name, "groups", &remainder)) {
+		if (len != sizeof(ctx[0].mc_group)) {
+			return -EINVAL;
+		}
+
+		group_id = remainder[0] - '0';
+		if (group_id < 0 || group_id > ARRAY_SIZE(ctx)) {
+			return -EINVAL;
+		}
+
+		rc = read_cb(cb_arg, &ctx[group_id].mc_group, len);
+		if (rc >= 0) {
+			return 0;
+		}
+
+		return rc;
+	}
+
+	return -ENOENT;
+}
+
+static struct settings_handler settings_conf = {
+	.name = MULTICAST_SETTINGS_BASE,
+	.h_set = multicast_settings_load,
+};
+
+static void multicast_session_start(struct k_work *work)
+{
+	int ret;
+
+	ret = lorawan_services_class_c_start();
+	if (ret < 0) {
+		LOG_WRN("Failed to switch to class C: %d. Retrying in 1s.", ret);
+		lorawan_services_reschedule_work(k_work_delayable_from_work(work), K_SECONDS(1));
+	}
+}
+
+static void multicast_session_stop(struct k_work *work)
+{
+	int ret;
+
+	ret = lorawan_services_class_c_stop();
+	if (ret < 0) {
+		LOG_WRN("Failed to revert to class A: %d. Retrying in 1s.", ret);
+		lorawan_services_reschedule_work(k_work_delayable_from_work(work), K_SECONDS(1));
+	}
+}
+
+static inline LoRaMacStatus_t multicast_group_setup(uint8_t id)
+{
+	__ASSERT_NO_MSG(id < ARRAY_SIZE(ctx));
+
+	McChannelParams_t channel = {
+		.IsRemotelySetup = true,
+		.IsEnabled = true,
+		.GroupID = (AddressIdentifier_t)id,
+		.Address = ctx[id].mc_group.addr,
+		.McKeys.McKeyE = ctx[id].mc_group.key_encrypted,
+		.FCountMin = ctx[id].mc_group.fcnt_min,
+		.FCountMax = ctx[id].mc_group.fcnt_max,
+		.RxParams = {0}
+	};
+
+	return LoRaMacMcChannelSetup(&channel);
+}
+
+static void multicast_package_callback(uint8_t port, bool data_pending, int16_t rssi, int8_t snr,
+				       uint8_t len, const uint8_t *rx_buf)
+{
+	uint8_t tx_buf[3 * MAX_MULTICAST_ANS_LEN];
+	uint8_t tx_pos = 0;
+	uint8_t rx_pos = 0;
+
+	__ASSERT(port == LORAWAN_PORT_MULTICAST, "Wrong port %d", port);
+
+	while (rx_pos < len) {
+		uint8_t command_id = rx_buf[rx_pos++];
+
+		if (sizeof(tx_buf) - tx_pos < MAX_MULTICAST_ANS_LEN) {
+			LOG_ERR("insufficient tx_buf size, some requests discarded");
```

Comment:
Under what conditions can this occur?

Reply:
In theory we can have any number of batch requests inside one LoRaWAN message (up to the size of the LoRaWAN payload). However, I have never observed that in real life (using AWS and ChirpStack backends). Added a comment to make it a bit more clear.